### PR TITLE
added account_alias in the response of module aws_caller_facts

### PR DESCRIPTION
**Pull request by sdubrul**
##### SUMMARY
Added an account_alias field in the output of the aws_caller_facts module.
I believe this fits nicely with the other information that is returned by this module and is a bit more  verbal in explaining what the used aws account actually is.

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
aws_caller_facts

##### ANSIBLE VERSION
```
ansible 2.7.0.dev0 (aws_caller_facts_alias 66dea8c065) last updated 2018/07/05 11:50:44 (GMT +200)
  config file = None
  configured module search path = [u'/Users/sdubrul/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/sdubrul/code/github/ansible/lib/ansible
  executable location = /Users/sdubrul/code/github/ansible/bin/ansible
  python version = 2.7.15 (default, Jun 17 2018, 12:46:58) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]
```

##### ADDITIONAL INFORMATION
before:
```
$ansible localhost -m aws_caller_facts -v

localhost | SUCCESS => {
    "account": "133256760XXX",
    "arn": "arn:aws:sts::1332567XXXXX:assumed-role/admin/sdubrul@xxxx.com",
    "changed": false,
    "user_id": "AROAJYFVM4AKNXXXXXXXX:sdubrul@xxxx.com"
}
```

after: 
```
$ansible localhost -m aws_caller_facts -v

localhost | SUCCESS => {
    "account": "1332567XXXXX",
    "account_alias": "cisco-prodops-dev",
    "arn": "arn:aws:sts::1332567XXXXX:assumed-role/admin/sdubrul@xxxx.com",
    "changed": false,
    "user_id": "AROAJYFVM4AKNXXXXXXXX:sdubrul@xxxx.com"
}

```

### CODE REVIEW

**Comment by sdubrul** [comment]
Shippable failed due to missing permissions:

`ClientError: An error occurred (AccessDenied) when calling the ListAccountAliases operation: User: arn:aws:sts::966509639900:assumed-role/ansible-core-ci-test-prod/prod=shippable=ansible=ansible=72735.66 is not authorized to perform: iam:ListAccountAliases on resource: *`

I'll let @orthanc comment on how to proceed further (i.e. fix the test or something else..)

**Comment by orthanc** [comment]
We'll have to wait on the core committers to add the permission to the shippable role, but that aside...

This looks like a good addition to me, and fits nicely with the intent of caller_facts.

Couple more things that do need to be added though:
* Can you extend the [integration test for this module](https://github.com/ansible/ansible/blob/devel/test/integration/targets/aws_caller_facts/tasks/main.yaml) to check for the new `account_alias` key in the response
* Can you add the new `iam:ListAccountAliases` permission to [security-policy.json](https://github.com/ansible/ansible/blob/devel/hacking/aws_config/testing_policies/security-policy.json) so that someone setting up the limited priveldges for the test can still use this module (also provides a good hint to the core team about the permissions missing from the integration build). It would be good to check that the integration test passes locally with the permission limited role or iam user once you've made this update. See [Integration Tests in the dev guide](https://docs.ansible.com/ansible/2.5/dev_guide/testing_integration.html) for info on how to run the tests and setup the iam permissions
* Can you add a changelog fragment for this as a minor change [here](https://github.com/ansible/ansible/tree/devel/changelogs/fragments)

I would be interested as well if you could explain a bit more abound how you see account alias being used. Is this just in debugging / logging or are you finding it useful programatically as well?

File: lib/ansible/modules/cloud/amazon/aws_caller_facts.py
Comment (orthanc) [inline lib/ansible/modules/cloud/amazon/aws_caller_facts.py]: Can you add a brief comment here to the effect of "There is only ever zero or one account alises" with link to AWS API doc on this.

I initially assumed this would behave incorrectly if there was more than one alias and it would be good to spare other's that confusion in future.

Thanks

File: lib/ansible/modules/cloud/amazon/aws_caller_facts.py
Comment (orthanc) [inline lib/ansible/modules/cloud/amazon/aws_caller_facts.py]: I think it would be best to rename this variable to caller_facts as we're now extending it's meaning to be more than strictly the caller identity.

I like the approach you've taken of just adding the account alias information into the existing dict, but just think we need to rename it since it's purpose is changed by this commit.

**Comment by sdubrul** [comment]
Thanks for reviewing as well as providing all the pointers @orthanc!
Your comments are all very helpful to polish this change.
I believe I added all your suggestions in the meantime too.

Regarding use cases you were spot on: logging or debugging the account name as this means more to an operator than the account id. Definitely when managing multiple accounts..
I guess you could even use the alias in the same way as the id to assert that plays are executed against the right cloud.

P.S. The test of this module worked fine as well on my env:

```
$test/runner/ansible-test integration -v aws_caller_facts
Using existing aws cloud config: test/integration/cloud-config-aws.yml
Running aws_caller_facts integration test role
Run command: ansible-playbook aws_caller_facts-sf9p2s.yml -i inventory -e @integration_config.yml -v -e @/Users/sdubrul/code/github/an ...
Using /Users/sdubrul/code/github/ansible/test/integration/integration.cfg as config file
 [WARNING]: The `junit_xml` python module is not installed. Disabling the `junit` callback plugin.


PLAY [amazon] *****************************************************************************************************************************

TASK [Gathering Facts] ********************************************************************************************************************
ok: [localhost]

TASK [aws_caller_facts : retrieve caller facts] *******************************************************************************************
ok: [localhost] => {"account": "13325676XXXX", "account_alias": "cisco-prodops-dev", "arn": "arn:aws:iam::13325676XXXX:user/ansible_test", "changed": false, "user_id": "AIDAJB4IN5VIS7HEXXXX"}

TASK [aws_caller_facts : assert correct keys are returned] ********************************************************************************
ok: [localhost] => {
    "changed": false,
    "msg": "All assertions passed"
}

PLAY RECAP ********************************************************************************************************************************
localhost                  : ok=3    changed=0    unreachable=0    failed=0

```

**Comment by orthanc** [comment]
shipit

Looks good, we'll need to wait for the core team to update the CI build permissions.

File: lib/ansible/modules/cloud/amazon/aws_caller_facts.py
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/aws_caller_facts.py:92]: I think if this fails, we can safely return `None` or `''` as the account alias, because it's possible people don't have the additional ListAccountAliases permission. That would be backwards compatible as well, since setups that already work without that permission would suddenly break in this version.

**Review by ryansb — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/amazon/aws_caller_facts.py
Comment (orthanc) [reply lib/ansible/modules/cloud/amazon/aws_caller_facts.py:92]: I like the idea of retaining backwards compatability on the permissons, but I'd like to maintain a distinction between being unable to get the account alias and there being no account alias.

How about the following:

* If we cannot get the account alias because of permissions of other failure: Do not include the account_alias key in the response
* If we are able to make the call and get the result of no account alias: Include account_alise = None in the response
* If we succeed in getting the account alias: Include account_alias = _the alias_ in the response

Comment (sdubrul) [reply lib/ansible/modules/cloud/amazon/aws_caller_facts.py:92]: I'm fine with the latest suggestion.
I would like to add a pointer in the documentation on this behavior too.
And possibly raise a warning?
What about the test case btw, how should that one behave?

Comment (orthanc) [reply lib/ansible/modules/cloud/amazon/aws_caller_facts.py:92]: In terms of the test, I would leave it as is. I think we get more value from testing the return in the case where this works rather than trying to simulate / ensure the permission failure.

And yes, please do include this in the documentation that the permission is required for that value to be returned.

Personally I'd say don't add a warning for this case. In my mind warnings should always be indicating something that you should change / fix. We're effectively taking the line that this is an optional addition and it's perfectly fine to use the module without this permission (as long as you don't need the alias). Assuming someone is expecting to use the alias then it's will fail pretty obviously so I don't know that adding a warning really helps in this case. this

Comment (ryansb) [reply lib/ansible/modules/cloud/amazon/aws_caller_facts.py:92]: I'd be ok with skipping an integration test for this, and adding the CI perms so the happy path (when the module does have permissions) is tested. Agreed w/ @orthanc about not adding a warning.

**Review by ryansb — APPROVED** [review]

**Comment by ryansb** [comment]
Oh, minor whitespace issue, but overall LGTM.

**Comment by orthanc** [comment]
Looks good.  shipit

**Comment by sdubrul** [comment]
bot_status

**Comment by sdubrul** [comment]
Anything else I can do to get this merged upstream? Cheers

**Comment by s-hertel** [comment]
Thanks @sdubrul!

**Pull request merged**
